### ISSUE TITLE
Maven dependencies not resolved by IDE

### ISSUE THREAD

**Issue by kmoughon** [codenvy/codenvy#351]
Originally from here: https://groups.google.com/a/codenvy.com/forum/#!topic/codenvy/panZOL0tjUk 

Original Issue:
I'm having an issue with Maven projects in Codenvy 4.4 On Prem. I am able to build the project successfully with Maven, however the IDE does not resolve the dependencies defined in the pom.xml and reports them in error. The IDE is able to resolve the classes in my src folder and works correctly for syntax highlighting and code completion etc. As a sanity check I loaded one of the sample java projects with external dependencies, the web-java-spring project in the same workspace, and it has the same issues.

Per suggestions on google groups, i tried adding the proxy to .bashrc. Latest status: 
If I am understanding correctly , not only does Maven need internet access to download the dependencies but the Codenvy Maven plugin does as well to download the files independently. I have tried adding the proxy to ~/.bashrc and it had no effect. Is there any action i need to take to make the Codenvy Maven plugin 'refresh' or retry to pick up the new settings? Restarting the workspace didn't work and cleared out the .bashrc file.

**Comment by TylerJewell** [comment]
@kmoughon - Hi, thanks for the detailed issue.

First, there is a separation between the proxy settings for the workspace itself (which are set within a Codenvy configuration property), and those used by maven.  You *may* need to have both set for maven to work properly.

Second, let's get your Codenvy proxy settings configured the right way.  You do not want to modify them in the .bashrc as those changes will not be persisted.  Your workspace is under management of Codenvy, so you want the proxying to be consistent for every workspace that is there.  When you first install Codenvy there is an --http-proxy-for-codenvy-workspaces parameter.  The value of this will be injected into the .bashrc of the workspace.  There is a matching --https-proxy and --no-proxy for this configuration item.  If your install is already done, you can change the system's values for these parameters using the Codenvy CLI.  See the "codenvy config <property> <value>" to change the value and then have the properties reapplied to the system.  You can type "codenvy config" to list all of the properties and find the right one for this configuration.

Third, once this is working, then your workspace has access to the internet through a proxy.  You may need to separately configure your maven repositories to use a proxy.  This is a limitation of maven itself, and it would be something that you add into the pom.xml of the maven repo.  There are many links online about how proxies are setup for maven to respect.

let us know how it goes.

**Comment by kmoughon** [comment]
@TylerJewell  thanks for the detailed explanation of the settings distribution.
I have verified that the proxy settings are configured as you said through the codenvy config CLI method, and this was done at install time. I have modified the ~/.m2/settings.xml file in the workspace to allow maven to use the proxy, and can successfully build the jar for my project. I am still having an issue where the IDE does not recognize that the dependencies are downloaded and reports errors. I can reproduce this in the default java-centos workspace (with only the settings.xml modification) with the web-java-spring example project. please see attached screenshots for verification. 
![examplebuildsuccess](https://cloud.githubusercontent.com/assets/10276740/16596153/a3f38fb6-42c0-11e6-8466-53bc05cc30fa.PNG)
![examplesettings](https://cloud.githubusercontent.com/assets/10276740/16596195/cdedf892-42c0-11e6-9b76-00508b153553.PNG)

**Comment by TylerJewell** [comment]
Ok.  If the jars are downloaded but you have errors like this then the issue is related to classpath of your workspace and not proxies.  What is the project type that has been set?  If it is the javac project type them in the project configuration you can specify the jar and source folders which will remove those issues!

**Comment by kmoughon** [comment]
The project type is Maven. We were able to make it work on another javac project by manually specifying the jars, but so far have been unsuccessful with Maven projects. I tried the sample Maven project in order to see if it was something wrong with my project or with the overall setup. Since the issue occurs in the sample project this points to either a system setup issue or an issue in Codenvy.

 The issue seems to be isolated to the IDE classpath. Since we set the project to type Maven, we cannot edit the classpath, the buttons are disabled in the popup window, shown under Project->configure classpath. Maven downloads the defined jars into ~/.m2/repository/ directory. I am not sure how Codenvy resolves this directory. Another interesting thing we have noticed is that dependencies defined in Maven do not get resolved however java utility classes like java.util.Date are correctly resolved by the IDE. You can see this on line 8 in the included screenshot. 
![correctlyresolve](https://cloud.githubusercontent.com/assets/10276740/16619176/1418e90e-435b-11e6-962f-eb8f370d4599.PNG)
![classpath](https://cloud.githubusercontent.com/assets/10276740/16619179/17033494-435b-11e6-868d-1a253871faf0.PNG)

**Comment by TylerJewell** [comment]
@dmytro-ndp @eivantsov @evidolob - any ideas?  Really whacky - they are behind a proxy.  Proxy configuration is confirmed to be correct and that the proxy is setup properly in the workspace itself.  Maven project type, but none of the JARs are being properly resolved in 4.4.

@kmoughon - can you export http_proxy and https_proxy environment variables from within the workspace itself?

**Comment by kmoughon** [comment]
I did a printenv in the workspace and can see both the http_proxy and https_proxy variables are set correctly. Internet access works in the workspace for wget and curl and yum.

**Comment by eivantsov** [comment]
@kmoughon What do you see when you expand external libraries in the project tree? Do you see external deps there?

**Comment by kmoughon** [comment]
There are some jars there, but none of the external ones appear.
![externaljars](https://cloud.githubusercontent.com/assets/10276740/16653365/64f02ed2-441e-11e6-9ccc-f9004e691f55.PNG)

**Comment by eivantsov** [comment]
@kmoughon Thanks for sharing a lot of details. 

I am going to run some tests and let you know what I have found out.

**Comment by eivantsov** [comment]
@kmoughon I have run some tests and here're my findings:

1. When you add your custom settings.xml, add it to `~/.m2` While Maven respects `/conf/settings.xml` location, Che Maven plugin does not.
2. Another solution would be to run build manually just like you did, and then re-import the project. Right click on the project tree to call a context menu, Maven > Reimport. As all of the deps will already be in a local Maven repo, Che maven plugin will find jars and easily place them into a classpath.

This is a bug on the plugin side though.

**Comment by kmoughon** [comment]
@eivantsov 
Sounds like you found the smoking gun. Could you elaborate on your #1, we do currently add the file in the workspace to ~/.m2/settings.xml. I am not familiar with the /conf/settings.xml location. We are using the centos jdk 8 built in workspace image if that matters for these directories. 

We tried your workaround described in #2 but could not get it to work on the web-java-spring demo project or any of our projects. 

Do we need to submit a bug report to Eclipse Che or will this be handled by someone at Codenvy?
thanks again for the help.

**Comment by eivantsov** [comment]
@kmoughon I think I have found the root cause - https://github.com/codenvy/dockerfiles/commit/0b861dd01fcc71f4cc9f4614241f5dd5d067a69d

While we require Maven 3.3.9 for the Maven plugin to work. 

I have rebuilt centos image. Can you give it another try? Make sure you pull the latest image though

**Comment by thaspius** [comment]
@eivantsov I'm working with Kirby and we've verified that this fixed the demo project/workspace. We are currently testing our custom environment.

**Comment by eivantsov** [comment]
@thaspius let me know if you have further questions.

**Comment by kmoughon** [comment]
@thaspius & @eivantsov 
I have confirmed that this fix also works in our custom environment. Appreciate the help from everyone.

**Comment by Yakushima** [comment]
On beta.codenvy.com, this helped me get the java-simple-console example going with Junit 4 rather than the (supplied) Junit 3 -- the workaround first mentioned by eivantsov on July 11th did the trick. But where should I report this against the beta? Very new to all this, so a pointer would be appreciated.

**Comment by TylerJewell** [comment]
Hi!  You are reporting in a good location.  Is there another issue you are uncovering?  

By the way - this pull request appeared last week for eclipse Che.  It is a robust unit test capability.  Once it is merged into Che it will appear in Codenvy.

https://github.com/eclipse/che/pull/2157

**Comment by Yakushima** [comment]
I suspect it's the exact same issue, just showing up in a simpler example: an almost-nothing console app, plus JUnit. Changing the pom to JUnit 4 seemed to get Maven to fetch the library, but the build still assumed the supplied JUnit 3 library. Anyway, very grateful to find this way to get unstuck -- as a noob, I just assumed the problem was me.
